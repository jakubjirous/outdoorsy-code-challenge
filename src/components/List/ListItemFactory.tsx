import React from "react";
import styled from "styled-components";
import { Rental } from "../../redux/features/rental/rentalApi";

const S = {
  Container: styled.div`
    display: flex;
    flex-flow: row nowrap;
    justify-content: flex-start;
    align-content: center;
    gap: var(--spacing-5);
  `,
  Name: styled.div`
    flex: 1;
    align-self: center;
    font-size: var(--fs-md);
    line-height: var(--lh-md);
  `,
};

interface Props {
  rental: Rental;
}

// TODO: image rendering (Jakub Jirous 2022-05-10 16:15:30)
// TODO: unit test (Jakub Jirous 2022-05-10 16:16:20)
function ListItemFactory({ rental }: Props) {
  const { id, name, imageUrl } = rental;

  return (
    <S.Container key={id}>
      Image
      <S.Name>{name}</S.Name>
    </S.Container>
  );
}

export default ListItemFactory;
