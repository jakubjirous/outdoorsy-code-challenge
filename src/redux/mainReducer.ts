import { combineReducers } from "@reduxjs/toolkit";
import { rentalApi } from "./features/rental/rentalApi";

export const mainReducer = combineReducers({
  // Add the generated reducer as a specific top-level slice
  [rentalApi.reducerPath]: rentalApi.reducer,
});
