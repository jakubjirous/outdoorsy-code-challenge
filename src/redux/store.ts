import { configureStore } from "@reduxjs/toolkit";
import { rentalApi } from "./features/rental/rentalApi";
import { mainReducer } from "./mainReducer";

export const store = configureStore({
  reducer: mainReducer,

  // Adding the api middleware enables caching, invalidation, polling, and other useful features of `rtk-query`.
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(rentalApi.middleware),
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
