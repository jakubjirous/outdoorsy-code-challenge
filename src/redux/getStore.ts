import { configureStore, StoreEnhancer } from "@reduxjs/toolkit";
import { rentalApi } from "./features/rental/rentalApi";
import { mainReducer } from "./mainReducer";
import { RootState, store } from "./store";

/**
 * Gets a store set up with some givens initial state.
 * The store is set up with the RTK Query middleware.
 * Useful mainly for creating redux store provider when writing unit tests.
 *
 * @param initialState
 * @param enhancers
 */
export function getStore(initialState?: RootState, enhancers?: StoreEnhancer[]): typeof store {
  return configureStore({
    reducer: mainReducer,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(rentalApi.middleware),
    enhancers,
  });
}
