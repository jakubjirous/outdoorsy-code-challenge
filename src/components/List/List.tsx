import React from "react";
import { useSelector } from "react-redux";
import AutoSizer, { Size } from "react-virtualized-auto-sizer";
import { FixedSizeList, ListChildComponentProps } from "react-window";
import styled from "styled-components";
import { useGetRentalQuery } from "../../redux/features/rental/rentalApi";
import { RootState } from "../../redux/store";
import ListItemFactory from "./ListItemFactory";

export const LIST_ITEM_SIZE = 166;

const S = {
  Section: styled.section`
    height: 100%;
  `,
};

// TODO: loading (Jakub Jirous 2022-05-10 16:15:30)
// TODO: no result (Jakub Jirous 2022-05-10 16:15:30)
// TODO: unit test (Jakub Jirous 2022-05-10 16:16:20)
function List() {
  const keywords = useSelector((state: RootState) => state.search.keywords);

  // Using a query hook automatically fetches data and returns query values
  const { data: rentals, error, isLoading } = useGetRentalQuery(keywords);

  const listItems = rentals?.map((rental) => ListItemFactory({ rental }));

  const Row = React.memo((props: ListChildComponentProps) => {
    const { data, index, style } = props;
    return <div style={style}>{data.items[index]}</div>;
  });

  return (
    <S.Section>
      {!rentals && isLoading ? (
        <>Loading...</>
      ) : rentals?.length === 0 || error ? (
        <>No results ...</>
      ) : (
        <AutoSizer>
          {(size: Size) => (
            <FixedSizeList
              itemSize={LIST_ITEM_SIZE}
              width={size.width}
              height={size.height}
              itemCount={listItems?.length ?? 0}
              itemData={{ items: listItems }}
            >
              {Row}
            </FixedSizeList>
          )}
        </AutoSizer>
      )}
    </S.Section>
  );
}

export default List;
