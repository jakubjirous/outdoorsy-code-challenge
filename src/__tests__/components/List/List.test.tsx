import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import List from "../../../components/List/List";
import { getStore } from "../../../redux/getStore";

const renderContainer = () => {
  return render(
    <Provider store={getStore()}>
      <List />
    </Provider>
  );
};

describe("<List/>", () => {
  test("should render correctly", () => {
    const { baseElement } = renderContainer();

    expect(baseElement).toMatchSnapshot();
  });
});
