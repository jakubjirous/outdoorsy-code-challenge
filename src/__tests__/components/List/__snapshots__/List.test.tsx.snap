// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List/> should render correctly 1`] = `
<body>
  <div>
    <section
      class="sc-papXJ hUHDjE"
    >
      <div
        class="sc-jSMfEi beGjEx"
      >
        <div
          class="sc-gKXOVf gnvyZk"
        >
          <div
            class="sc-bczRLJ kkcSEP"
            height="8.4375rem"
            width="11.25rem"
          />
          <div
            class="sc-bczRLJ fNpSxG"
            height="var(--fs-lg)"
            width="50%"
          />
        </div>
        <div
          class="sc-gKXOVf gnvyZk"
        >
          <div
            class="sc-bczRLJ kkcSEP"
            height="8.4375rem"
            width="11.25rem"
          />
          <div
            class="sc-bczRLJ fNpSxG"
            height="var(--fs-lg)"
            width="50%"
          />
        </div>
        <div
          class="sc-gKXOVf gnvyZk"
        >
          <div
            class="sc-bczRLJ kkcSEP"
            height="8.4375rem"
            width="11.25rem"
          />
          <div
            class="sc-bczRLJ fNpSxG"
            height="var(--fs-lg)"
            width="50%"
          />
        </div>
        <div
          class="sc-gKXOVf gnvyZk"
        >
          <div
            class="sc-bczRLJ kkcSEP"
            height="8.4375rem"
            width="11.25rem"
          />
          <div
            class="sc-bczRLJ fNpSxG"
            height="var(--fs-lg)"
            width="50%"
          />
        </div>
      </div>
    </section>
  </div>
</body>
`;
